lib/powerpoint.js

```diff
@@ -61,81 +72,88 @@ class Powerpoint extends EventEmitter2 {
+      return new Promise((resolve) => {
+        const gsCmd = gsCmdRoot + ` -o ${imgDir}/img-${pdfIndex}-%d.png ${pdfPath}`
+        exec(gsCmd, (err, stdout, stderr) => {
+          this.emit('done.gs.convert', { command: gsCmd, time: this.elapsed(start), error: err })
+          if (err) {
+            gsErr.push(err)
+          }
+          resolve()
+        })
+      })
+    })
 
-  _convert (converter, outputDir, done) {
-    pdfLogger('converting pdf into images...')
-    var self = this
-    var start = this.nowInMillis()
-    var imgStart = start
-    converter.pdf.convert((err, imagePath) => {
-      if (err) {
-        self.emit('err.png.single', {output: imagePath, error: err, time: this.elapsed(imgStart)})
-        errorLogger(err)
+    // GS executes each PDF asynchronously, so we need a collection of promises
+    // to wait for all image files to be present
+    Promise.all(requests).then(() => {
+      if (!_.isEmpty(gsErr)) {
+        this.emit('err.png.all', {error: gsErr, time: this.elapsed(start)})
       } else {
-        self.emit('done.png.single', {output: imagePath, time: this.elapsed(imgStart)})
-      }
-      imgStart = this.nowInMillis()
-      // Will it be faster to add images here?
-    }, (err, images) => {
-      if (err) {
-        done(err)
-        self.emit('err.png.all', {output: images, error: err, time: this.elapsed(start)})
-      } else {
-        self.emit('done.png.all', {output: images, time: this.elapsed(start)})
+        const images = fs.readdirSync(imgDir).map(f => `${imgDir}/${f}`)
+        this.emit('done.png.all', {output: images, time: this.elapsed(start)})
         var pptxOutput = path.resolve(outputDir, `output_${process.hrtime()[1]}.pptx`)
-        this._aggregateSlides(images, pptxOutput, done, outputDir)
+        this._aggregateSlides(images, pptxOutput, callback, outputDir)
       }
     })
   }
 
+  _getConvertOptions (options) {
+    const co = options.convertOptions || {}
+    const o = {}
+    /* Note: if the density is too low and there is a slide with a transparent background,
+     The image may show a horizontal line on the slide when it is rendered in the PPTX.
+     (was visible at 72, but not visible at 150)
+     */
+    o.density = co.density ? Math.min(co.density, DPI_DENSITY_MAX) : DPI_DENSITY_DEFAULT
+    return o
+  }
+
+  /**
+   *
+   * @param {array} images
+   * @param pptxOutput pptx file path
+   * @param done callback
+   * @param outputDir the directory where png and pptx files are generated
+   * @private
+   */
   _aggregateSlides (images, pptxOutput, done, outputDir) {
     this._createSlides(images, pptxOutput, (slideErr, output) => {
-      pdfLogger('Finished rendering all slides')
       done(slideErr, output)
-      rmdir(path.resolve(outputDir, 'img'),
-        (e) => e && errorLogger('Could not delete working directory:', e))
+      this.clean && rmdir(path.resolve(outputDir, 'img'), (e) => e && errorLogger('Could not delete working directory:', e))
```

Comment:
This line is kind of hard to follow because it's so long.  I think it would be better split onto multiple lines.